core/src/main/scala/akka/persistence/cassandra/query/EventsByPersistenceIdStage.scala

```diff
@@ -116,42 +119,76 @@ import scala.util.{ Failure, Success, Try }
 
     type Extractor[T <: Extracted] = (Row, EventDeserializer, Serialization) => T
 
+    final case class SeqNrValue(sequenceNr: Long) extends Extracted
+
+    final case class PersistentReprValue(persistentRepr: PersistentRepr) extends Extracted {
+      override def sequenceNr: Long = persistentRepr.sequenceNr
+    }
+
+    val persistentRepr: Extractor[PersistentReprValue] = (row, ed, s) => {
+      val persistentRepr = extractPersistentRepr(row, ed, s)
+      PersistentReprValue(persistentRepr)
+    }
+
     val taggedPersistentRepr: Extractor[TaggedPersistentRepr] = (row, ed, s) => {
-      val tags = extractTags(row)
+      val tags = extractTags(row, ed)
+      val persistentRepr = extractPersistentRepr(row, ed, s)
+      TaggedPersistentRepr(persistentRepr, tags, row.getUUID("timestamp"))
+    }
+
+    val optionalTaggedPersistentRepr: Extractor[OptionalTagged] = (row, ed, s) => {
+      val seqNr = row.getLong("sequence_nr")
+      val tags = extractTags(row, ed)
+      if (tags.isEmpty) {
+        // no tags, no need to extract more
+        OptionalTagged(seqNr, OptionVal.None)
+      } else {
+        val persistentRepr = extractPersistentRepr(row, ed, s)
+        val tagged = TaggedPersistentRepr(persistentRepr, tags, row.getUUID("timestamp"))
+        OptionalTagged(seqNr, OptionVal.Some(tagged))
+      }
+    }
+
+    val sequenceNumber: Extractor[SeqNrValue] = (row, ed, s) => {
```

Comment:
Nice, we could make this even better by having a different query so we don't pull back all the data like we do for the higest sequence nr query. Another time!